mava/networks/utils/sable/decode.py

```diff
@@ -0,0 +1,169 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Tuple
+
+import chex
+import distrax
+import jax
+import jax.numpy as jnp
+from flax import linen as nn
+
+# General shapes legend:
+# B: batch size
+# S: sequence length
+# A: number of actions
+# N: number of agents
+
+
+def train_decoder_fn(
+    decoder: nn.Module,
+    obs_rep: chex.Array,
+    action: chex.Array,
+    legal_actions: chex.Array,
+    hstates: chex.Array,
+    dones: chex.Array,
+    step_count: chex.Array,
+    n_agents: int,
+    chunk_size: int,
+    rng_key: Optional[chex.PRNGKey] = None,
+) -> Tuple[chex.Array, chex.Array]:
+    """Parallel action sampling for discrete action spaces."""
+    # Delete `rng_key` since it is not used in discrete action space
+    del rng_key
+
+    shifted_actions = get_shifted_actions(action, legal_actions, n_agents=n_agents)
+
+    logit, _ = act_decoder_fn(
+        decoder=decoder,
+        obs_rep=obs_rep,
+        shifted_actions=shifted_actions,
+        hstates=hstates,
+        dones=dones,
+        step_count=step_count,
+        legal_actions=legal_actions,
+        chunk_size=chunk_size,
+    )
+
+    masked_logits = jnp.where(
+        legal_actions,
+        logit,
+        jnp.finfo(jnp.float32).min,
+    )
+
+    distribution = distrax.Categorical(logits=masked_logits)
+    action_log_prob = distribution.log_prob(action)
+    action_log_prob = jnp.expand_dims(action_log_prob, axis=-1)
+    entropy = jnp.expand_dims(distribution.entropy(), axis=-1)
+
+    return action_log_prob, entropy
+
+
+def act_decoder_fn(
+    decoder: nn.Module,
+    obs_rep: chex.Array,
+    shifted_actions: chex.Array,
+    hstates: Tuple[chex.Array, chex.Array],
+    dones: chex.Array,
+    step_count: chex.Array,
+    legal_actions: chex.Array,
+    chunk_size: int,
+) -> Tuple[chex.Array, Tuple[chex.Array, chex.Array]]:
+    logit = jnp.zeros_like(legal_actions, dtype=jnp.float32)
+
+    # Apply the decoder per chunk
+    num_chunks = shifted_actions.shape[1] // chunk_size
+    for chunk_id in range(0, num_chunks):
+        start_idx = chunk_id * chunk_size
+        end_idx = (chunk_id + 1) * chunk_size
+        chunked_obs_rep = obs_rep[:, start_idx:end_idx]
+        chunk_shifted_actions = shifted_actions[:, start_idx:end_idx]
+        chunk_dones = dones[:, start_idx:end_idx]
+        chunk_step_count = step_count[:, start_idx:end_idx]
+        chunk_logit, hstates = decoder(
```

Comment:
I know I've been removing a lot of your comments, but I think this will help a bit as it's quite a big block of code 
```suggestion
        start_idx = chunk_id * chunk_size
        end_idx = (chunk_id + 1) * chunk_size
        # chunk obs, actions, dones and step counts
        chunked_obs_rep = obs_rep[:, start_idx:end_idx]
        chunk_shifted_actions = shifted_actions[:, start_idx:end_idx]
        chunk_dones = dones[:, start_idx:end_idx]
        chunk_step_count = step_count[:, start_idx:end_idx]
        
        chunk_logit, hstates = decoder(
```